File: easy/LongestCommonPrefix.java
/**
 * Completed: 2025-05-09
 * Difficulty: Easy
 * 14. Longest Common Prefix
 * https://leetcode.com/problems/longest-common-prefix/description/
 */

class LongestCommonPrefix {
    public String longestCommonPrefix(String[] strs) {
        String prefix = strs[0];

        for (int i = 1; i < strs.length; i++) {
            String curWord = strs[i];
            int endPointer = prefix.length() - 1;
            boolean prefixMatch = false;
            if (prefix.length() > curWord.length()) {
                prefix = prefix.substring(0, curWord.length());
            }
            while (!prefixMatch && endPointer >= 0) {
                if (curWord.substring(0, prefix.length()).equals(prefix)) {
                    prefixMatch = true;
                } else {
                    prefix = prefix.substring(0, prefix.length() - 1);
                }
                endPointer--;
            };
        }
        return prefix;
    }
}

File: easy/longest_common_prefix.py
class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        """
        Understand: 

        longest common prefix

        Match:
        String prefix
        Loop

        Pseudo Code:

        idx = 0
        max_prefix = false

        

        while not max_prefix and idx < len(strs[0]):
            prefix = strs[0].substring(0,idx)
            for str in strs:
                if prefix not in str:
                    max_prefix = true
            idx += 1

        Implement
        """

        idx = 0
        prefix = None

        while idx < len(strs[0]):
            idx += 1
            prefix = strs[0][0:idx]
            for string in strs:
                if not string.startswith(prefix):
                    idx -= 1
                    break

        return strs[0][0:idx]

        """
        Review:

        ["dog", "racecar", "car"]

        idx: 0
        mp: false
        prefix: ""

        idx: 1
        mp: false
        prefix: "f"

        idx: 2
        mp: false
        prefix: "fl"

        idx: 3
        mp: true
        prefix: "flo"

        """

